transaction/core/src/input_rules.rs

```diff
@@ -53,6 +69,103 @@ impl InputRules {
+            for fractional_output in self.fractional_outputs.iter() {
+                // Try to unblind the fractional output amount.
+                let fractional_output_amount = fractional_output.reveal_amount()?;
+                // Let's check if there is a corresponding real output.
+                let real_output = tx
+                    .prefix
+                    .outputs
+                    .iter()
+                    .find(|x| {
+                        x.public_key == fractional_output.tx_out.public_key
+                            && x.target_key == fractional_output.tx_out.target_key
+                    })
+                    .ok_or(InputRuleError::MissingRealOutput)?;
+                // Let's try to unblind its amount, using amount shared secret from the
+                // fractional output (which should be the same)
+                let real_output_amount = try_reveal_amount(
+                    real_output,
+                    fractional_output.amount_shared_secret.as_ref(),
+                )?;
+
+                // Check the bounds of the real change amount
+                if real_output_amount.token_id != fractional_output_amount.token_id {
+                    return Err(InputRuleError::RealOutputTokenIdMismatch);
+                }
+                if real_output_amount.value > fractional_output_amount.value {
+                    return Err(InputRuleError::RealOutputAmountExceedsFractional);
+                }
+
+                // Check that the fill fraction is respected.
+                // Intuitively, what we are doing is checking that:
+                //
+                // real_output_value >= n/d * fractional_output_value,
+                //
+                // where n/d is the fill fraction. The fill fraction is 1 if
+                // real_change_amount = 0, and proportionally less as more of
+                // the input is returned as change.
+                // However, to avoid numerical issues, we multiply this out and
+                // verify this as a comparison of u128 values:
+                //
+                // real_output_value * d >= n * fractional_output_value
+                if (real_output_amount.value as u128 * fill_fraction_denom)
+                    < (fractional_output_amount.value as u128 * fill_fraction_num)
+                {
+                    return Err(InputRuleError::RealOutputAmountDoesNotRespectFillFraction);
+                }
+            }
+        } else {
+            // If there is no fractional change output specified, then none of the related
+            // fractional output rules or rule verification data should be present.
+            if !self.fractional_outputs.is_empty() {
+                return Err(InputRuleError::FractionalOutputsNotExpected);
+            }
+            if self.max_allowed_change_value != 0 {
+                return Err(InputRuleError::MaxAllowedChangeValueNotExpected);
+            }
+        }
+
```

Comment:
I would rather not break this up to this extent, I think it makes it harder to follow and not easier.

Reply:
I think breaking up `verify()` to the three
```rust
        self.verify_max_tombstone_block(tx)?;
        self.verify_required_outputs(tx)?;
        self.verify_partial_fill_rules(tx)?;
        Ok(())
    }
```
Would be good.  As it currently sits it's relying on code comments to break up the larger groupings of what's being verified.  If a fourth one gets added it will likely be after the partial fill logic and be fairly easy to lose sight of